segment/filer_test.go

```diff
@@ -22,6 +22,39 @@ func TestFileName(t *testing.T) {
 	require.Equal(t, "00000007394872394732-00ab1234cd4567ef.wal", fn)
 }
 
+func TestSegmentReadBeforeAfterFirstIndex(t *testing.T) {
```

Comment:
nit: I think the core feature we need to test that caused this bug was reading after a non-first log is the first entry in the segment i.e. the first Log's Index != BaseIndex.

How about:

```suggestion
func TestSegmentStartingFromHighIndexThanBase(t *testing.T) {
```

Or something?